Olá, 

Meu nome é Ramon. Ao ao longo do texto farei algumas observações sobre melhorias no código. Nas partes em que não houver comentários, está tudo certo, ok? Estarei aberto a feedbacks e discussões sobre o tema.

**Peço que mantenha e não altere os comentários que eu fizer por aqui para que possamos nos localizar posteriormente, ok?**

Mais uma coisa, vamos utilizar um código de cores para você entender os meus feedbacks no seu notebook. Funciona assim:

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Sucesso. Tudo foi feito corretamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Alerta não crítco, mas que pode ser corrigido para melhoria geral no seu código/análise
</div>

<div class="alert alert-block alert-danger">

<b>Comentário do revisor: </b> <a class="tocSkip"></a>
    
Erro que precisa ser consertado, caso contrário seu projeto **não** será aceito.
</div>

Você pode me responder usando isso:
<div class="alert alert-block alert-info">
<b>Resposta do Aluno.</b> 
    Bom dia Ramon, tudo bom? Quando enviei o projeto já havia ficado 4 horas tentando calibrar hiperparâmetros sem encontrar nenhuma melhoria significativa nos resultados. Você tem alguma sugestão de como fazer esta calibragem se forma mais eficiente?<a class="tocSkip"></a>
</div>

###

<div class="alert alert-block alert-success">
<b>Comentário Geral do Revisor</b> <a class="tocSkip"></a>

Obrigado por enviar seu projeto. 
    
Parabéns, você fez um bom trabalho.

Como pontos fortes do projeto, destaco:
- Código simples e eficiente (sem loop desnecessário)
- Uso adequado dos recursos da lição

    
<span class="text-danger">Porém, há espaço para melhorias conforme os itens apontados anteriormente. Destaco os  pontos a seguir: </span>
 * <span class="text-danger"> Precisa calibrar hiperparâmetros </span>


Portanto, sugiro que você corrija os pontos indicados.
    
Se você tiver alguma dúvida sobre de algo que você não tenha entendido, sinta-se à vontade para perguntar ao tutor ou deixar um comentário em **azul** para que, na próxima iteração, eu possa responder.

Conte conosco. Lembre-se de que você pode tirar dúvidas com revisores usando os comentários ou buscar ajuda de um tutor.
Bom trabalho. 
</div>




<div class="alert alert-block alert-success">
<b> Comentário Geral do revisor v2: </b> <a class="tocSkip"></a>

O projeto voltou sem calibração de hiperparâmetros. Você pode usar o GridSearchCV nos parâmetros depth para random forest e árvore decisão apenas. Não precisa calibrar lightgbm. Oriento apenas alterar a parte referente a esses dois modelos.</div>

<div class="alert alert-block alert-success">
<b> Comentário Geral do revisor v3: </b> <a class="tocSkip"></a>

Tudo certo agora. Desejo sucesso na jornada. Parabéns pelo empenho.
</div>

# Quer andar de carro velho, amor?

O serviço de vendas de carros usados Rusty Bargain está desenvolvendo um aplicativo para atrair novos clientes. Nesse aplicativo, você pode descobrir rapidamente o valor de mercado do seu carro. Você tem acesso a dados históricos: especificações técnicas, versões de acabamento e preços. Você precisa construir o modelo para determinar o valor. 

Rusty Bargain está interessado em:

- a qualidade da predição;
- a velocidade da predição;
- o tempo necessário para o treinamento

## Preparação de Dados

In [3]:
# Carregando Bibliotecas
import time

import pandas as pd
import numpy as np
import lightgbm as lgb

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import GridSearchCV

In [4]:
# Carregando os dados
data = pd.read_csv("car_data.csv")

df = pd.DataFrame(data)

In [5]:
# Visualizando os dados
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 354369 entries, 0 to 354368
Data columns (total 16 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   DateCrawled        354369 non-null  object
 1   Price              354369 non-null  int64 
 2   VehicleType        316879 non-null  object
 3   RegistrationYear   354369 non-null  int64 
 4   Gearbox            334536 non-null  object
 5   Power              354369 non-null  int64 
 6   Model              334664 non-null  object
 7   Mileage            354369 non-null  int64 
 8   RegistrationMonth  354369 non-null  int64 
 9   FuelType           321474 non-null  object
 10  Brand              354369 non-null  object
 11  NotRepaired        283215 non-null  object
 12  DateCreated        354369 non-null  object
 13  NumberOfPictures   354369 non-null  int64 
 14  PostalCode         354369 non-null  int64 
 15  LastSeen           354369 non-null  object
dtypes: int64(7), object(

In [6]:
df.sample(5, random_state=333)

DateCrawled  Price VehicleType  RegistrationYear Gearbox  Power  \
308816  30/03/2016 14:45    800         NaN              2016  manual     75   
60879   31/03/2016 17:53      0       coupe              2001    auto    306   
262934  26/03/2016 19:51   1100       wagon              1999  manual    101   
11910   01/04/2016 19:49    400       small              2002  manual     60   
2472    01/04/2016 22:57   3300       wagon              2004  manual    131   

         Model  Mileage  RegistrationMonth  FuelType          Brand  \
308816   civic   150000                  9       NaN          honda   
60879       cl   150000                  7    petrol  mercedes_benz   
262934  passat   150000                 11    petrol     volkswagen   
11910    punto   150000                  0       NaN           fiat   
2472    passat   150000                 10  gasoline     volkswagen   

       NotRepaired       DateCreated  NumberOfPictures  PostalCode  \
308816         NaN  20/03/2016 00:00                 0       30966   
60879           no  31/03/2016 00:00                 0       92637   
262934          no  26/03/2016 00:00                 0       25594   
11910          yes  01/04/2016 00:00                 0       40225   
2472           yes  01/04/2016 00:00                 0       53498   

                LastSeen  
308816  06/04/2016 12:16  
60879   06/04/2016 11:44  
262934  28/03/2016 09:16  
11910   05/04/2016 16:45  
2472    02/04/2016 00:41

In [7]:
# Conferindo se há valores extremos
df.describe()

Price  RegistrationYear          Power        Mileage  \
count  354369.000000     354369.000000  354369.000000  354369.000000   
mean     4416.656776       2004.234448     110.094337  128211.172535   
std      4514.158514         90.227958     189.850405   37905.341530   
min         0.000000       1000.000000       0.000000    5000.000000   
25%      1050.000000       1999.000000      69.000000  125000.000000   
50%      2700.000000       2003.000000     105.000000  150000.000000   
75%      6400.000000       2008.000000     143.000000  150000.000000   
max     20000.000000       9999.000000   20000.000000  150000.000000   

       RegistrationMonth  NumberOfPictures     PostalCode  
count      354369.000000          354369.0  354369.000000  
mean            5.714645               0.0   50508.689087  
std             3.726421               0.0   25783.096248  
min             0.000000               0.0    1067.000000  
25%             3.000000               0.0   30165.000000  
50%             6.000000               0.0   49413.000000  
75%             9.000000               0.0   71083.000000  
max            12.000000               0.0   99998.000000

In [8]:
# Analisando valores incomuns
df["RegistrationYear"].sort_values().unique()

array([1000, 1001, 1039, 1111, 1200, 1234, 1253, 1255, 1300, 1400, 1500,
       1600, 1602, 1688, 1800, 1910, 1915, 1919, 1920, 1923, 1925, 1927,
       1928, 1929, 1930, 1931, 1932, 1933, 1934, 1935, 1936, 1937, 1938,
       1940, 1941, 1942, 1943, 1944, 1945, 1946, 1947, 1948, 1949, 1950,
       1951, 1952, 1953, 1954, 1955, 1956, 1957, 1958, 1959, 1960, 1961,
       1962, 1963, 1964, 1965, 1966, 1967, 1968, 1969, 1970, 1971, 1972,
       1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 1982, 1983,
       1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994,
       1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005,
       2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016,
       2017, 2018, 2019, 2066, 2200, 2222, 2290, 2500, 2800, 2900, 3000,
       3200, 3500, 3700, 3800, 4000, 4100, 4500, 4800, 5000, 5300, 5555,
       5600, 5900, 5911, 6000, 6500, 7000, 7100, 7500, 7800, 8000, 8200,
       8455, 8500, 8888, 9000, 9229, 9450, 9996, 99

In [9]:
df["Power"].sort_values().unique()

array([    0,     1,     2,     3,     4,     5,     6,     7,     8,
           9,    10,    11,    12,    13,    14,    15,    16,    17,
          18,    19,    20,    21,    22,    23,    24,    25,    26,
          27,    28,    29,    30,    31,    32,    33,    34,    35,
          36,    37,    38,    39,    40,    41,    42,    43,    44,
          45,    46,    47,    48,    49,    50,    51,    52,    53,
          54,    55,    56,    57,    58,    59,    60,    61,    62,
          63,    64,    65,    66,    67,    68,    69,    70,    71,
          72,    73,    74,    75,    76,    77,    78,    79,    80,
          81,    82,    83,    84,    85,    86,    87,    88,    89,
          90,    91,    92,    93,    94,    95,    96,    97,    98,
          99,   100,   101,   102,   103,   104,   105,   106,   107,
         108,   109,   110,   111,   112,   113,   114,   115,   116,
         117,   118,   119,   120,   121,   122,   123,   124,   125,
         126,   127,

O conjunto de dados possui 16 colunas, sendo 15 características e 1 objetivo. São elas:

- **DateCrawled** - Data em que o perfil foi baixado do banco de dados;
 - Formato: *object*, deveria ser *DateTime* porém não será convertido por não conter dados relevantes para o modelo e que podem confundir o modelo afinal.
 - Não possui valores nulos.
- **Price** - Objetivo de descoberta do algoritmo *(target)*;
 - Formato: *int*, está correto.
 - Não possui valores nulos.
 - Aparentemente sem valores extremos.
- **VehicleType** - Tipo de carroceria do veículo *(feature)*;
 - Formato: *object*, correto.
 - Possui valores nulos que podem ser substituídos por *unknown*.
- **RegistrationYear** - Ano de matrícula do veículo *(feature)*;
 - Formato: *int*, está correto.
 - Não possui valores nulos.
 - Possui valores extremos e datas futuras, possivelmente por conta de erros de digitação, serão substituídos por 2024 *(ano atual)*;
- **Gearbox** - Tipo de caixa de transmissão *(feature)*;
 - Formato: *object*, correto.
 - Possui valores nulos que podem ser substituídos por *unknown*.
- **Power** - Potência (hp) *(feature)*;
 - Formato: *int*, correto.
 - Não possui valores nulos.
 - Aparentemente sem valores extremos.
- **Model** - Modelo do veículo *(feature)*;
 - Formato: *object*, correto.
 - Possui valores nulos que podem ser substituídos por *unknown*.
- **Mileage** - Quilomeatragem (km) *(feature)*;
 - Formato: *int*, correto.
 - Não possui valores nulos.
 - Aparentemente sem valores extremos.
- **RegistrationMonth** - Mês de registro do veículo *(feature)*;
 - Formato: *int*, correto.
 - Não possui valores nulos.
 - Aparentemente sem valores extremos.
- **FuelType** - Tipo de combustível *(feature)*;
 - Formato: *object*, correto.
 - Possui valores nulos que podem ser substituídos por *unknown*.
- **Brand** - Marca do veículo *(feature)*;
 - Formato: *object*, correto.
 - Não possui valores nulos.
- **NotRepaired** - Veículo reparado ou não *(feature)*;
 - Formato: *object*, também pode ser convertido para *boolean* para melhor interpretação do modelo.
 - Possui valores nulos que podem ser substituídos por *unknown*, neste caso deixa de ser *boolean*.
- **DateCreated** - Data de criação do perfil *(feature)*;
 - Formato: *object*, precisa ser convertido para *Datetime*.
 - Não possui valores nulos.
- **NumberOfPictures** - Número de fotos do veículo *(feature)*;
 - Formato: *int*, correto.
 - Não possui valores nulos.
 - Sem valores extremos *(apesar de só possui valores zero ainda pode ser uma coluna a ser levada em consideração em futuras predições)*.
- **PostalCode** - Código postal do proprietário do veículo *(feature)*;
 - Formato: *int*, correto.
 - Não possui valores nulos.
 - Aparentemente sem valores extremos *(valores categóricos)*.
- **LastSeen** - Data da última atividade do usuário *(feature)*.
 - Formato: *object*, precisa ser convertido para *Datetime*.
 - Não possui valores nulos.

In [10]:
# Retirando a coluna 'DateCrawled'
df = df.drop(["DateCrawled"], axis=1)

In [11]:
# Substituindo valores nulos por 'unknown'
df.fillna("unknown", inplace=True)


<div class="alert alert-block alert-warning">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Você está substituindo todos os valores nans por unknown. Precisa tomar cuidado para não misturar os tipos dos dados
</div>


In [12]:
# Convertendo 'DateCreated' para datetime
df["DateCreated"] = pd.to_datetime(df["DateCreated"], format="%d/%m/%Y %H:%M")

In [13]:
# Convertendo 'LastSeen' para datetime
df["LastSeen"] = pd.to_datetime(df["LastSeen"], format="%d/%m/%Y %H:%M")

In [14]:
# Substituindo anos de registro maiores que 2024 na coluna 'RegistrationYear'
df["RegistrationYear"] = (
    df["RegistrationYear"].apply(lambda x: 2024 if x > 2024 else x).astype("int")
)

## Treinamento do modelo

In [15]:
# Separando conjuntos de treinamento e de teste
df_train, df_valid = train_test_split(df, test_size=0.25, random_state=12345)

In [16]:
# Separando features de target
train_features = df_train.drop(["Price"], axis=1)
valid_features = df_valid.drop(["Price"], axis=1)
train_target = df_train["Price"]
valid_target = df_valid["Price"]

In [17]:
# Fazendo codificação de rótulos
train_features_encoded = train_features.copy()
valid_features_encoded = valid_features.copy()

In [18]:
cols = ["VehicleType", "Gearbox", "Model", "FuelType", "Brand", "NotRepaired"]

for col in cols:
    train_features_encoded[col] = LabelEncoder().fit_transform(
        train_features_encoded[col]
    )
    valid_features_encoded[col] = LabelEncoder().fit_transform(
        valid_features_encoded[col]
    )

In [19]:
# Fazendo codificação de datas
dfs = [train_features_encoded, valid_features_encoded]

for df in dfs:
    df["CreatedDay"] = df["DateCreated"].dt.day
    df["CreatedMonth"] = df["DateCreated"].dt.month
    df["CreatedYear"] = df["DateCreated"].dt.year
    df["LastSeenDay"] = df["LastSeen"].dt.day
    df["LastSeenMonth"] = df["LastSeen"].dt.month
    df["LastSeenYear"] = df["LastSeen"].dt.year
    df["LastSeenHour"] = df["LastSeen"].dt.hour

train_features_encoded = train_features_encoded.drop(
    ["LastSeen", "DateCreated"], axis=1
)
valid_features_encoded = valid_features_encoded.drop(
    ["LastSeen", "DateCreated"], axis=1
)

In [20]:
train_features_encoded.info()

<class 'pandas.core.frame.DataFrame'>
Index: 265776 entries, 25945 to 217570
Data columns (total 19 columns):
 #   Column             Non-Null Count   Dtype
---  ------             --------------   -----
 0   VehicleType        265776 non-null  int64
 1   RegistrationYear   265776 non-null  int64
 2   Gearbox            265776 non-null  int64
 3   Power              265776 non-null  int64
 4   Model              265776 non-null  int64
 5   Mileage            265776 non-null  int64
 6   RegistrationMonth  265776 non-null  int64
 7   FuelType           265776 non-null  int64
 8   Brand              265776 non-null  int64
 9   NotRepaired        265776 non-null  int64
 10  NumberOfPictures   265776 non-null  int64
 11  PostalCode         265776 non-null  int64
 12  CreatedDay         265776 non-null  int32
 13  CreatedMonth       265776 non-null  int32
 14  CreatedYear        265776 non-null  int32
 15  LastSeenDay        265776 non-null  int32
 16  LastSeenMonth      265776 non-null  int

### Regressão Linear

In [21]:
%time
# Criando modelo de Regressão Linear para Prova Real dos outros modelos
LRmodel = LinearRegression()

LRmodel.fit(train_features_encoded, train_target)

CPU times: total: 0 ns
Wall time: 0 ns


LinearRegression()

- Tempo de treinamento para modelo de Regressão Linear: **5.96 µs**.

### Gradient Boosting

In [22]:
# Convertendo os dados para o formato Dataset
train_data = lgb.Dataset(train_features_encoded, label=train_target)
valid_data = lgb.Dataset(
    valid_features_encoded, label=valid_target, reference=train_data
)

In [23]:
# Definindo parâmetros
params = {
    "learning_rate": 0.1,
    "seed": 12345,
    "verbosity": 2,
    "num_leaves": 2000,
    "min_data_in_leaf": 100,
    "min_data_per_group": 2000,
}

In [24]:
# Medindo o tempo de treinamento
start_time = time.time()

# Treinando o modelo
n_iters = 100
lgb_model = lgb.train(params, train_data, n_iters, valid_sets=[valid_data])

# Calculando e exibindo o tempo de treinamento
end_time = time.time()
training_time = end_time - start_time
print(f"Tempo de treinamento: {training_time:.2f} segundos")

[LightGBM] [Debug] Dataset::GetMultiBinFromSparseFeatures: sparse rate 0.764165
[LightGBM] [Debug] Dataset::GetMultiBinFromAllFeatures: sparse rate 0.164773
[LightGBM] [Debug] init for col-wise cost 0.016147 seconds, init for row-wise cost 0.050676 seconds
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.068950 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 1034
[LightGBM] [Info] Number of data points in the train set: 265776, number of used features: 16
[LightGBM] [Info] Start training from score 4413.365319
[LightGBM] [Debug] Trained a tree with leaves = 2000 and depth = 25
[LightGBM] [Debug] Trained a tree with leaves = 2000 and depth = 24
[LightGBM] [Debug] Trained a tree with leaves = 2000 and depth = 27
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Debug] Trained a tree with leaves = 1998 and depth = 25
[LightGBM] [Debug] Trained a tree with leaves = 20

- Tempo de treinamento para modelo de Gradient Boosting: **309.19 s**.

### Floresta de Regressão

In [25]:
# Treinando o modelo de Floresta de Regressão
RFmodel_1 = RandomForestRegressor(
    random_state=12345, warm_start=True, max_features="sqrt"
)

In [ ]:
# Calibrando hiperparâmetros
parameters = {
    "n_estimators": range(5, 16, 5),
    "max_depth": range(25, 126, 100),
    "min_samples_split": range(70, 771, 700),
    "min_samples_leaf": range(50, 551, 500),
    "max_leaf_nodes": range(100, 1001, 900),
}

clf = GridSearchCV(RFmodel_1, parameters, verbose=3, scoring="neg_mean_squared_error")
clf.fit(train_features_encoded, train_target)

Fitting 5 folds for each of 48 candidates, totalling 240 fits
[CV 1/5] END max_depth=25, max_leaf_nodes=100, min_samples_leaf=50, min_samples_split=70, n_estimators=5;, score=-5706153.429 total time=   2.5s
[CV 2/5] END max_depth=25, max_leaf_nodes=100, min_samples_leaf=50, min_samples_split=70, n_estimators=5;, score=-5706874.162 total time=   2.8s
[CV 3/5] END max_depth=25, max_leaf_nodes=100, min_samples_leaf=50, min_samples_split=70, n_estimators=5;, score=-5757108.211 total time=   1.9s
[CV 4/5] END max_depth=25, max_leaf_nodes=100, min_samples_leaf=50, min_samples_split=70, n_estimators=5;, score=-5516969.405 total time=   1.7s
[CV 5/5] END max_depth=25, max_leaf_nodes=100, min_samples_leaf=50, min_samples_split=70, n_estimators=5;, score=-5886508.924 total time=   2.0s
[CV 1/5] END max_depth=25, max_leaf_nodes=100, min_samples_leaf=50, min_samples_split=70, n_estimators=10;, score=-5506108.028 total time=   3.3s
[CV 2/5] END max_depth=25, max_leaf_nodes=100, min_samples_leaf=50,

In [25]:
clf.best_estimator_

RandomForestRegressor(max_depth=25, max_features='sqrt', max_leaf_nodes=1000,
                      min_samples_leaf=50, min_samples_split=70,
                      n_estimators=15, random_state=12345, warm_start=True)

In [42]:
%time
# Treinando o modelo de Floresta de Regressão com hiperparâmetros selecionados
RFmodel = RandomForestRegressor(
    max_depth=25,
    max_features='sqrt',
    max_leaf_nodes=1000,
    min_samples_leaf=50,
    min_samples_split=70,
    n_estimators=15,
    random_state=12345,
    warm_start=True
)
RFmodel.fit(train_features_encoded, train_target)

CPU times: user 2 µs, sys: 0 ns, total: 2 µs
Wall time: 4.53 µs


RandomForestRegressor(max_depth=25, max_features='sqrt', max_leaf_nodes=1000,
                      min_samples_leaf=50, min_samples_split=70,
                      n_estimators=15, random_state=12345, warm_start=True)

- Tempo de treinamento para modelo de Floresta Aleatória de Regressão: **4.53 µs**.


<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Precisa calibrar hiperparâmetro
</div>


### Árvore de Regressão

In [27]:
# Treinando o modelo de Árvore de Regressão
DTmodel_1 = DecisionTreeRegressor(random_state=12345)

In [28]:
# Calibrando hiperparâmetros
parameters = {
    "max_depth": range(25, 126, 100),
    "min_samples_split": range(70, 771, 700),
    "min_samples_leaf": range(50, 551, 500),
    "max_leaf_nodes": range(1000, 3001, 1000),
}

clf = GridSearchCV(DTmodel_1, parameters, verbose=3, scoring="neg_mean_squared_error")
clf.fit(train_features_encoded, train_target)

Fitting 5 folds for each of 24 candidates, totalling 120 fits
[CV 1/5] END max_depth=25, max_leaf_nodes=1000, min_samples_leaf=50, min_samples_split=70; total time=   1.1s
[CV 2/5] END max_depth=25, max_leaf_nodes=1000, min_samples_leaf=50, min_samples_split=70; total time=   1.0s
[CV 3/5] END max_depth=25, max_leaf_nodes=1000, min_samples_leaf=50, min_samples_split=70; total time=   1.0s
[CV 4/5] END max_depth=25, max_leaf_nodes=1000, min_samples_leaf=50, min_samples_split=70; total time=   1.0s
[CV 5/5] END max_depth=25, max_leaf_nodes=1000, min_samples_leaf=50, min_samples_split=70; total time=   1.1s
[CV 1/5] END max_depth=25, max_leaf_nodes=1000, min_samples_leaf=50, min_samples_split=770; total time=   1.1s
[CV 2/5] END max_depth=25, max_leaf_nodes=1000, min_samples_leaf=50, min_samples_split=770; total time=   1.0s
[CV 3/5] END max_depth=25, max_leaf_nodes=1000, min_samples_leaf=50, min_samples_split=770; total time=   1.0s
[CV 4/5] END max_depth=25, max_leaf_nodes=1000, min_sam

[CV 5/5] END max_depth=125, max_leaf_nodes=1000, min_samples_leaf=550, min_samples_split=70; total time=   0.9s
[CV 1/5] END max_depth=125, max_leaf_nodes=1000, min_samples_leaf=550, min_samples_split=770; total time=   0.9s
[CV 2/5] END max_depth=125, max_leaf_nodes=1000, min_samples_leaf=550, min_samples_split=770; total time=   0.9s
[CV 3/5] END max_depth=125, max_leaf_nodes=1000, min_samples_leaf=550, min_samples_split=770; total time=   0.9s
[CV 4/5] END max_depth=125, max_leaf_nodes=1000, min_samples_leaf=550, min_samples_split=770; total time=   0.9s
[CV 5/5] END max_depth=125, max_leaf_nodes=1000, min_samples_leaf=550, min_samples_split=770; total time=   0.9s
[CV 1/5] END max_depth=125, max_leaf_nodes=2000, min_samples_leaf=50, min_samples_split=70; total time=   1.2s
[CV 2/5] END max_depth=125, max_leaf_nodes=2000, min_samples_leaf=50, min_samples_split=70; total time=   1.2s
[CV 3/5] END max_depth=125, max_leaf_nodes=2000, min_samples_leaf=50, min_samples_split=70; total tim

GridSearchCV(estimator=DecisionTreeRegressor(random_state=12345),
             param_grid={'max_depth': range(25, 126, 100),
                         'max_leaf_nodes': range(1000, 3001, 1000),
                         'min_samples_leaf': range(50, 551, 500),
                         'min_samples_split': range(70, 771, 700)},
             scoring='neg_mean_squared_error', verbose=3)

In [29]:
clf.best_estimator_

DecisionTreeRegressor(max_depth=125, max_leaf_nodes=3000, min_samples_leaf=50,
                      min_samples_split=70, random_state=12345)

In [43]:
%time
# Treinando o modelo de Árvore de Regressão com hiperparâmetros selecionados
DTmodel = DecisionTreeRegressor(
    max_depth=125,
    max_leaf_nodes=3000,
    min_samples_leaf=50,
    min_samples_split=70,
    random_state=12345
)
DTmodel.fit(train_features_encoded, train_target)

CPU times: user 3 µs, sys: 0 ns, total: 3 µs
Wall time: 4.05 µs


DecisionTreeRegressor(max_depth=125, max_leaf_nodes=3000, min_samples_leaf=50,
                      min_samples_split=70, random_state=12345)

- Tempo de treinamento para modelo de Árvore de Regressão: **4.05 µs**.

Sobre os tempos de treinamento, o modelo mais rápido é o de *Árvore de Regressão* com tempo de execução de **4.05 µs**.

Ranking de tempo de treinamento dos modelos:
- **1º - Árvore de Regressão** - 4.05 µs
- **2º - Floresta de Regressão** - 4.53 µs
- **3º - Regressão Linear** - 5.96 µs
- **4º - Gradient Boosting** - 309.19 s


<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Precisa calibrar hiperparâmetro
</div>


## Análise do modelo

### Análise de Qualidade

Será usada a métrica **REQM** para avaliar a qualidade dos diferentes modelos.

In [38]:
# Avaliando modelo de Regressão Linear
LR_pred = LRmodel.predict(valid_features_encoded)
LR_mse = mean_squared_error(valid_target, LR_pred)
print("REQM do modelo de Regressão Linear:", LR_mse)

REQM do modelo de Regressão Linear: 12455713.567873783


In [32]:
# Avaliando modelo de Gradient Boosting
lgb_pred = lgb_model.predict(valid_features_encoded)
lgb_mse = mean_squared_error(valid_target, lgb_pred)
print("REQM do modelo de Gradient Boosting:", lgb_mse)

REQM do modelo de Gradient Boosting: 2991702.960958839


In [33]:
# Avaliando modelo de Floresta de Regressão
RF_pred = RFmodel.predict(valid_features_encoded)
RF_mse = mean_squared_error(valid_target, RF_pred)
print("REQM do modelo de Floresta de Regressão:", RF_mse)

REQM do modelo de Floresta de Regressão: 4339839.477974582


In [34]:
# Avaliando modelo de Árvore de Regressão
DT_pred = DTmodel.predict(valid_features_encoded)
DT_mse = mean_squared_error(valid_target, DT_pred)
print("REQM do modelo de Árvore de Regressão:", DT_mse)

REQM do modelo de Árvore de Regressão: 4101568.175017456


O modelo com melhor REQM é o *Gradient Boosting*, tendo a métrica quase **30%** melhor do que os outros 2 melhores modelos!

Ranking de REQM dos modelos:
- **1º - Gradient Boosting** - 2.991.702,96
- **2º - Árvore de Regressão** - 4.101.568,17
- **3º - Floresta de Regressão** - 4.339.839,47
- **4º - Regressão Linear** - 12.455.713,56

### Análise de velocidade de execução

In [39]:
%%timeit

# Avaliando modelo de Regressão Linear
LR_pred = LRmodel.predict(valid_features_encoded)

14.9 ms ± 3.88 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [35]:
%%timeit

# Avaliando modelo de Gradient Boosting
lgb_pred = lgb_model.predict(valid_features_encoded)

3.93 s ± 40.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [36]:
%%timeit

# Avaliando modelo de Floresta de Regressão
RF_pred = RFmodel.predict(valid_features_encoded)

105 ms ± 529 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [37]:
%%timeit

# Avaliando modelo de Árvore de Regressão
DT_pred = DTmodel.predict(valid_features_encoded)

11.7 ms ± 52.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


Sobre as velocidades de execução, o modelo mais rápido é o de *Árvore de Regressão* com tempo de execução de **52.8 µs**.

Ranking de velocidades de execução dos modelos:
- **1º - Árvore de Regressão** - 52.8 µs
- **2º - Floresta de Regressão** - 529 µs
- **3º - Regressão Linear** - 3.88 ms
- **4º - Gradient Boosting** - 40.8 ms

## Conclusão

Este trabalho explorou a aplicação de técnicas de aprendizado de máquina para predição de valores de mercado de veículos para o serviço de venda de carros usados, "Rusty Bargain". O objetivo foi analisar o impacto de diferentes modelos com Árvores de Decisão contra um modelo básico de Regressão Linear nos seguintes quesitos:

- Qualidade da predição;
- Velocidade da predição;
- Tempo necessário para o treinamento.

### Sobre os Dados

- Os dados foram convertidos para os tipos mais adequados, os dados ausentes foram todos preenchidos por *'unknown'* por se tratarem todos de serem categorias em colunas contendo somente *'strings'*.

- Foi feita a Codificação de Rótulos e Datas conforme *ano/mês/dia* para correta identificação dos modelos.

### Comparação de Modelos

#### Qualidade da Predição

- **Melhor Modelo:** *Gradient Boosting* apresentou o menor REQM (Erro Quadrático Médio Relativo).
- **REQM:** 2.991.702,96
- **Desempenho:** 30% melhor que os outros modelos.

**Ranking de Qualidade:**
- Gradient Boosting - 2.991.702,96
- Árvore de Regressão - 4.101.568,17
- Floresta de Regressão - 4.339.839,47
- Regressão Linear - 12.455.713,56

#### Velocidade de Predição
- **Mais Rápido:** *Árvore de Regressão* com tempo de execução de 52.8 µs.

**Ranking de Velocidade:**
- Árvore de Regressão - 52.8 µs
- Floresta de Regressão - 529 µs
- Regressão Linear - 3.88 ms
- Gradient Boosting - 40.8 ms

#### Tempo de Treinamento
- **Mais Rápido:** *Árvore de Regressão* com tempo de execução de 4.05 µs.

**Ranking de Tempo de Treinamento:**
- Árvore de Regressão - 4.05 µs
- Floresta de Regressão - 4.53 µs
- Regressão Linear - 5.96 µs
- Gradient Boosting - 309.19 s

### Considerações:

- O estudo demonstrou que, apesar de mais lentos, os algoritmos de *Gradient Boosting* são **muito** mais eficazes que outros modelos de regressão.
- Independente do modelo a ser utilizado o REQM ainda deixa muito a desejar...

### Recomendação

Considerando que, uma vez treinado, o modelo fará somente novas *predições* a cada acesso, a recomendação é utilizar o modelo de **Gradient Boosting** com os seguintes hiperparâmetros:

- `n_iters = 100`
- `learning_rate = 0.1`
- `seed = 12345`
- `num_leaves = 2000`
- `min_data_in_leaf = 100`
- `min_data_per_group = 2000`

# Checklist

Digite 'x' para verificar. Em seguida, pressione Shift + Enter.

- [x]  O Jupyter Notebook está aberto
- [x]  O código está livre de erros
- [x]  As células com o código foram organizadas em ordem de execução
- [x]  Os dados foram baixados e preparados
- [x]  Os modelos foram treinados
- [x]  A análise de velocidade e qualidade dos modelos foi realizada